Add vision for conda CLI blog post (#333)
conda/conda-dot-org: blog/2026-04-17-vision-for-conda-cli.mdx

---

> title: "Building a Better conda CLI: a Vision"

Reviewer: Minor: the lowercase `a` after the colon looks unintended given the rest of the title is title-cased.

```suggestion
title: "Building a Better conda CLI: A Vision"
```

slug: "2026-04-17-vision-for-conda-cli"
authors: [danyeaw]
tags: [conda, vision]
description: "A vision for the conda CLI in 2026: making conda fast, trusted, and delightful."
image: img/blog/2026-04-17-vision-for-conda-cli/banner.png
image_credit: <a href="https://unsplash.com/de/@mcnoble?utm_source=unsplash&utm_medium=referral&utm_content=creditCopyText">Matt Noble on Unsplash</a>
---

import useBaseUrl from "@docusaurus/useBaseUrl";
import ThemedImage from "@theme/ThemedImage";

> A vision for `conda` in 2026 is a package manager that is fast (real-time progress, sharded repodata, Rattler integration), trusted (better error messages, smarter confirmation friction, streaming output), and delightful (sensible defaults, helpful suggestions, intentional visuals, accessibility). Beyond the core CLI, the roadmap includes native PyPI wheel support, declarative `conda.toml` project environments, single-binary installers, and a comprehensive API for IDE and agent integrations.

Reviewer: The TL;DR's `fast` pillar currently names Rattler as the only technology, which makes the speed story read as rattler-only. The `libmamba` solver and a `conda-spawn`-style shell activation are already pulling weight here and you mention both further down, so I think it's fair to surface them up top too.

```suggestion
A vision for `conda` in 2026 is a package manager that is fast (real-time progress, sharded repodata, the `libmamba` solver, shell activation without startup cost, Rattler integration), trusted (better error messages, smarter confirmation friction, streaming output), and delightful (sensible defaults, helpful suggestions, intentional visuals, accessibility). Beyond the core CLI, the roadmap includes native PyPI wheel support, declarative `conda.toml` project environments, single-binary installers, and a comprehensive API for IDE and agent integrations.
```

Author: Hi @jezdez, I like the idea to add in shell integration and I added that. I don't think that `libmamba` solver is a key part of a future looking vision, so I removed that part. However, I did later on mention that we are building on a foundation of speed improvements that we already have in place.

Reviewer: Sounds good to me, that's a better example indeed!

- Sharded repodata – As I mentioned above, sharded repodata is a game changer for performance, and the next step is that we need to enable this by default and ensure it is available on a wider set of channels.
- Shell startup – There should be no shell startup penalty for using `conda`, and my vision is that a conda-spawn approach to spawn a new subshell will become the default `conda activate` and `conda deactivate` behavior.
- Rattler integration – We would like to use [Rattler](https://github.com/conda/rattler) more, a high-performance Rust implementation of core conda primitives. The performance floor rises significantly when the resolver and file I/O are written in Rust.

Our goal isn't to win a benchmark, but to make conda feel fast and responsive. Jaime Rodríguez-Guerra created a demo called [`conda-ng`](https://github.com/jaimergp/conda-ng/) that demonstrates what a blazingly fast `rattler`-based `conda` might look like:

<ThemedImage
  alt="Conda CLI in action"
  sources={{
    light: useBaseUrl("/img/blog/2026-04-17-vision-for-conda-cli/conda-ng-light.gif"),
    dark: useBaseUrl("/img/blog/2026-04-17-vision-for-conda-cli/conda-ng-dark.gif"),
  }}
/>

> <br />

Reviewer: Related to the TL;DR comment: the only demo in this section is `conda-ng`, which is rattler-based. A one-liner after the demo that acknowledges the Python-core speed work already in `conda` would keep the section from reading as rattler-only. Feel free to adjust the wording.

```suggestion
<br />

And this is not only a Rust story. The default `libmamba` solver and sharded repodata already make today's Python `conda` a lot faster, and deeper `rattler` integration builds on that work rather than replacing it. My vision is to keep advancing both together.
```

Author: Hi @jezdez, thanks for the suggestion. I think I captured the essence here with my own verbiage.

## Excellence Beyond the CLI

Fast, Trusted, and Delightful apply to more than the core CLI experience. A few other areas where this vision extends:

- Native PyPI support – If you look closely at the animation above, `conda` 1.4 shipped with a `conda pip` command. This shows that the Python ecosystem and PyPI are inextricably linked to `conda` from the very beginning. It is also a problem we have been trying to solve ever since. The [`conda-pypi`](https://github.com/conda/conda-pypi) plugin already supports natively installing wheels without shelling out to `pip`. My vision goes further: a conda channel that indexes all versions of all pure-Python wheels from PyPI directly in repodata, so they are always available without leaving the conda ecosystem.

> - Declarative environments – Modern development workflows need modern configuration. We have recently been working on improving cross-platform lockfile support, and we need to continue to build on that. Poetry, `uv`, and `pixi` have shown that software engineers really value having the ability to run tasks and manage their dependencies through projects or workspaces. I envision that we support these workflows with an ecosystem-standardized `conda.toml` configuration file to help make your projects easier to manage and more reproducible.

Reviewer: Small pointer so readers can connect this `conda.toml` vision to the `conda-workspaces` entry below.

```suggestion
- Declarative environments – Modern development workflows need modern configuration. We have recently been working on improving cross-platform lockfile support, and we need to continue to build on that. Poetry, `uv`, and `pixi` have shown that software engineers really value having the ability to run tasks and manage their dependencies through projects or workspaces. I envision that we support these workflows with an ecosystem-standardized `conda.toml` configuration file to help make your projects easier to manage and more reproducible, in the same direction `conda-workspaces` is already exploring.
```

[`conda-lockfiles`](https://github.com/conda-incubator/conda-lockfiles) is launching in the next few weeks in beta. It adds support for multiplatform lockfile formats, including `conda-lock` and `pixi.lock`, making cross-platform environment reproducibility a first-class `conda` feature.

[`conda-self`](https://github.com/conda-incubator/conda-self) is also launching in beta soon. It adds a `conda self` command that lets you safely manage your `base` environment, including protecting it from accidental modifications.

The following three projects are proof-of-concepts from Jannis Leidel that explore what future `conda` workflows could look like:

[`conda-workspaces`](https://github.com/conda-incubator/conda-workspaces) brings a pixi-like project-based workflow to `conda`, letting you manage multi-environment workspaces and tasks from a single configuration file.

[`conda-global`](https://github.com/conda-incubator/conda-global) enables installing CLI tools into isolated environments and placing them on your PATH, similar to `pipx install`, `uv tool`, or `pixi global`.

> [`conda-express`](https://github.com/jezdez/conda-express) is the lightweight, single-binary bootstrapper shown above. My vision is that it becomes the future Miniconda and Miniforge installer themselves, a small static binary that gets you a fully functional `conda` installation in seconds.

Reviewer: Two things bundled here:

1. Reorder so `conda-workspaces` comes last, and give it a vision-level statement that matches the `conda-express` one. Workspaces is the most integrative of the bunch (multi-env + features + task runner + `conda.lock` + pixi-manifest compatibility), and I'd like to highlight it as the future of environment management in `conda`.
2. Grounded the description in what the README actually says: `conda.toml` manifest (with `pixi.toml` / `pyproject.toml` compatibility), delegates solving to `conda` core, integrates with `conda-pypi` for PyPI deps. Also fixes the earlier `installer themselves` → `installer itself` agreement nit and echoes my earlier "cover all kinds of gaps" framing in the intro so the section hangs together.

```suggestion
The following three projects are proof-of-concepts from Jannis Leidel that explore what future `conda` workflows could look like. Together with `conda-lockfiles` and `conda-self`, they paint a picture of covering all kinds of gaps in `conda` today:

[`conda-global`](https://github.com/conda-incubator/conda-global) enables installing CLI tools into isolated environments and placing them on your PATH, similar to `pipx install`, `uv tool`, or `pixi global`.

[`conda-express`](https://github.com/jezdez/conda-express) is the lightweight, single-binary bootstrapper shown above. My vision is that it becomes the future Miniconda and Miniforge installer itself, a small static binary that gets you a fully functional `conda` installation in seconds.

[`conda-workspaces`](https://github.com/conda-incubator/conda-workspaces) brings a pixi-like project-based workflow to `conda` with project-scoped environments, composable features, a task runner with dependency ordering, and reproducible `conda.lock` files, all defined in a `conda.toml` (with `pixi.toml` and `pyproject.toml` manifests also supported). My vision is that workspaces become the future of environment management in `conda`. It builds on `conda` itself for solving and installation, and uses [`conda-pypi`](https://github.com/conda/conda-pypi) to resolve PyPI dependencies alongside conda packages. Cloning a repo should be all it takes to start working in a reproducible project.
```

> These projects prove the ideas are sound. But plugins can only go so far, and the next step is bringing these same improvements into `conda` itself.

Reviewer: The current framing reads as "plugins are scaffolding, conda-core is the destination." That doesn't match how the plugin system actually works: these projects will keep living as plugins (several in `conda-incubator`), and the pattern for making one first-class is the `conda-libmamba-solver` one, where `conda` takes it on as a dependency and ships it as the default without absorbing its code. Suggested rewrite to reflect that and to keep both tracks first-class.

```suggestion
These projects prove the ideas are sound, and they will keep evolving as plugins, which is exactly what `conda`'s plugin system is for. Several of them are incubating in [`conda-incubator`](https://github.com/conda-incubator) today. As these plugins mature, `conda` itself can take them on as dependencies and ship them as default behavior, the same way [`conda-libmamba-solver`](https://github.com/conda/conda-libmamba-solver) became the default solver. My vision is that we keep investing in both the plugin ecosystem and core `conda`.
```

Author: Hi @jezdez, thanks for the feedback. I shortened this up while trying to keep the essence that plugins and core conda enhancements are both needed.
